Board

nekr0z opened the pull request:
Примерно так мне увиделась концепция доски:

- состояние позволяет проверить валидность очередного хода, не обращаясь к истории партии;
- «наружу» торчат только методы, детали имплементации скрыты и могут быть изменены при необходимости;
- парсер FEN есть, сохранялку в FEN ~допишу~ написал — можно ~будет~ сохранять состояние доски в виде строки хоть в базу, хоть куда, опять таки, без зависимости от деталей имплементации; можно, конечно, написать сохранялку в любой другой формат.

В принципе, в общих чертах готово. Остались слабые места, но можно будет по ходу доработать.

sadmadrus committed:
Merge pull request #2 from sadmadrus/dev

Merge pull request #1 from sadmadrus/main

nekr0z committed:
WIP

nekr0z committed:
WIP

nekr0z committed:
Merge branch 'main' into board

sadmadrus commented on internal/board/board.go:
```diff
@@ -0,0 +1,231 @@
+// проверяется; на доске может оказаться 8 белых королей и чёрный ферзь,
+// держащий под боем каждого короля, или белая пешка на 1-й горизонтали.
+func FromFEN(fen string) (*Board, error) {
+	ss := strings.Split(fen, " ")
```
А разделитель разве не "/" по стандарту?

nekr0z replied:
> А разделитель разве не "/" по стандарту?

Это между горизонталями, отрабатывается далее по коду.
Тут разделитель между положением на доске и сервисными всякими штуками (чей ход, какие доступны рокировки и т.п.) — там по стандарту пробел.

Если что, я для теста из Википедии примеры FEN копировал :)

sadmadrus replied:
а... Что-то не увидел :)

nekr0z committed:
добавлен вывод доски в FEN

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,427 @@
+	b.blk = false
+}
+
+// BlackToMove устанавливает следующих ход белых.
```
опечатка - следующий ход черных

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,427 @@
+	default:
+		return nil, er
+	}
+	if ss[2] != "-" {
```
Наверное опечатка - мы разве не ss[3] должны проверять на enPassant? ss[2] это про рокировки же?

nekr0z replied:
Да!

Это бы поймалось в тестах потом, но у Вас глаз-алмаз!

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,427 @@
+)
+
+// Get возвращает фигуру, стоящую в заданной клетке.
+func (b *Board) Get(s square) (piece, error) {
```
не нужно проверять что s валидно (от 0 до 63)?

nekr0z replied:
А оно проверяется! :)

Тут идея в том, что тип `square` не экспортирован, и единственный способ для вызывающего передать значение `square` — использовать Sq():
```go
piece, err := b.Get(board.Sq("c5"))
piece, err := b.Get(board.Sq(17))
```
и т.п. Sq() возвращает `-1` для невалидных клеток, и Board.Get() выдаст ошибку.

loskutovanl replied:
точно, увидела

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,427 @@
+		b.fm++
+	}
+	b.blk = !b.blk
+	pc := b.brd[from]
```
правило 50 ходов: "если на протяжении последних 50 ходов каждого игрока ни одна [фигура] не была взята и ни одна [пешка] не сделала хода". я не понимаю, как b.brd[to]== 0 проверяет что фигура не была взята?

loskutovanl replied:
Пример. a2-a4 b4-a3 на а3 нет фигуры, но фигура была взята?

nekr0z replied:
На момент проверки brd[to] ещё содержит предыдущее (до хода) значение; если там не нуль, значит, на клетке, куда мы перемещаем фигуру, стояла фигура — очевидно, произошло взятие :)
Может, сделать проверку на цвет, чтобы своих не брать?

nekr0z replied:
> Пример. a2-a4 b4-a3 на а3 нет фигуры, но фигура была взята?

Пешка двигалась! :)

nekr0z committed:
исправлены опечатки

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,427 @@
+		b.RemoveCastling(CastlingWhiteQueenside)
+	case BlackPawn:
+		if from >= 8*6 && from < 8*7 && to >= 8*4 && to < 8*5 {
+			b.ep = 8*5 + to%8
```
немного проще для восприятия мне кажется будет b.ep = from - 8 для черных и b.ep = from + 8 для белых

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,427 @@
+		}
+	}
+	b.brd[from] = 0
+	b.brd[to] = pc
```
Если пешка достигла конечной диагонали, то b.brd[to] будет не pc, а другая фигура выбранная пользователем. Это уйдет в другой сервис?

nekr0z replied:
Пока не знаю, это надо будет решать. Как и с рокировками, и со взятием пешек на проходе.

nekr0z replied:
В общем — решено. В рамках метода Move она остаётся пешкой, но есть отдельный метод Promote(), который эту ситуацию обрабатывает.

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,427 @@
+
+// IsEnPassant возвращает true, если заданная клетка была перепрыгнута
+// пешкой в прошлом ходу.
+func (b *Board) IsEnPassant(s square) bool {
```
почему мы сравниваем клетку s с 0? нужно b.ep сравнивать?

nekr0z replied:
Сравнение с нулём — это во избежание ошибок, дальше там сравнение с b.ep.

Это — один из артефактов того, что пустой Board — валидная доска. Мы ж не хотим, чтоб для пустой доски `b.IsEnPassant(Sq("a1"))` возвращала `true`, а без этой проверки получится именно так.

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,427 @@
+
+// SetMoveNumber устанавливает номер хода (полного) на доске.
+func (b *Board) SetMoveNumber(n int) {
+	b.fm = n - 1
```
почему не b.fm = n ?

nekr0z replied:
Потому что для пустой доски ход первый, см. комментарий к этому полю в объявлении структуры Board.

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,427 @@
+			sb.WriteByte('/')
+		}
+	}
+	rr := make([]string, 16)
```
Я вообще не поняла конструкцию с rr , зачем она нужна и зачем мы в нее записываем минусы а потом замещаем их :(

nekr0z replied:
Для FEN надо вместо пустых клеток указывать их количество подряд. У нас в цикле пустые клетки рисуются как `"-"`.

Добавил комментарий, чтобы было понятнее, что происходит.

loskutovanl replied:
почитала про метод NewReplacer c примерами и стала понятнее

nekr0z committed:
добавлены лицензионные заголовки

nekr0z committed:
добавлено немного читаемости

nekr0z committed:
добавлена инициализация доски

nekr0z committed:
добавлены тесты

nekr0z committed:
добавлен тест

nekr0z committed:
добавлено fuzz-тестирование

nekr0z commented:
Начинаю ненавидеть фаззер — он находит всё новые и новые баги в парсере FEN :)

nekr0z committed:
fix bugs

nekr0z committed:
add UsFEN converter

nekr0z committed:
поддержка UsFEN

nekr0z committed:
Merge branch 'dev' into board

nekr0z committed:
clean up .gitignore

Пока нет внешнего фаззера, можно это не игнорировать. Если прикрутим
когда-нибудь, добавим.

nekr0z committed:
add logic to naming

nekr0z committed:
clean up .gitignore

nekr0z committed:
add license header

nekr0z commented on internal/board/board.go:
```diff
@@ -0,0 +1,521 @@
+	return b.brd[s], nil
+}
+
+// Move перемещает фигуру из from в to.
+// Валидность хода не проверяется, но по итогам ход переходит к
+// другому игроку, количество полуходов и ходов обновляется,
+// обновляются данные о пешке en passant и рокировках.
+// Чтобы сделать рокировку, надо, например, сделать ход королём,
+// а ладью убрать с доски и поставить на новое место, чтобы это
+// защиталось как только один ход.
+func (b *Board) Move(from, to square) error {
```
Мне самому этот метод не нравится, я его сделал больше ради тестов, и у меня чешутся руки его выпилить: без валидации он всё равно неполноценный, и, если по-хорошему, нарушает S в SOLID.
Может, ну его, а тесты я перепишу, чтоб он не нужен был? Как общественность?

sadmadrus replied:
а куда предлагаешь вынести?

nekr0z replied:
Сложный вопрос.
Может быть, действительно, есть смысл оставить здесь, но тогда надо доработать, чтобы работали взятие на проходе, рокировки и превращение пешек.
Я вот как мыслю: в том виде, в котором есть, этот метод бесполезен; по-хорошему, он бесполезен вообще — без валидации-то. И, коль скоро у нас есть сервис валидации, вот пусть там новое состояние доски и генерируется.
Но — было мне откровение сегодня поутру — _есть один момэнт_. У нас валидатор ходов — внешний микросервис, и нам надо как-то защититься от перехвата и подмены информации, чтобы читер не мог «валидировать» себе неправильный ход. И вот тут такой метод очень может помочь: сервис игровой сессии (или где там у нас сама игра происходит) использует этот метод, чтобы «увидеть» как будет выглядеть доска после хода, и сравнить с возвращённой от валидатора — так можно убедиться, что ход, который валидатор одобрил — это именно тот ход, о котором его спрашивали.
Правда, проблему перехвата с запретом оппоненту валидного хода это не решает.

sadmadrus replied:
Можно обойти эту проблему так: все данные от пользователя приходят всегда в сервис игровой сессии, а тот выступает роутером-аккумулятором данных. Получится, что для пользователя всего одна входная точка

nekr0z replied:
Нет, я говорю о случае, когда пользователь — злонамеренный хакер, который нашёл способ перехватывать и править запросы между сервисом сессии и валидатором ходов. Это ведь разные микросервисы, общаются они меж собой по сети, ergo, есть возможность перехвата.

nekr0z committed:
add en passant behavior

nekr0z committed:
add castling

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,607 @@
+			if b.ep >= 5*8 && b.brd[to-8] == BlackPawn {
+				b.brd[to-8] = 0
+			}
+		case BlackPawn:
```
не может появиться неожиданный сюрприз если b.ep = 0 в case BlackPawn?

например белая пешка а2, белая ладья а1. Черная пешка б2 съедает белую ладью а1. тогда Move удалит пешку на а2...

Может добавить в условие && b.ep > 0 ?

nekr0z replied:
Great catch!
Прочитал твоё описание ситуации, подумал: «Да ну, не может быть!» Написал тест, который это имитирует — и да, баг!
Исправляю.

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,607 @@
+		k = BlackKing
+	}
+
+	// под 960 эту логику надо будет переписать
```
Мысли вслух.
Напрашивается метод Get(p piece) square {}
которые по фигуре будет искать ее клетку на доске. И для короля такой понадобится при валидации, и для поиска ладьи в 960

nekr0z replied:
Это будет медленно. Над 960 я ещё думаю, там, конечно, придётся искать, но там можно ограничить поиск в худшем случае семью клетками, это сильно лучше, чем 64 для общего случая.

loskutovanl commented on internal/board/board.go:
```diff
@@ -0,0 +1,607 @@
+// Board представляет доску с позицией. Инициализация необязательна,
+// пустой Board готов к употреблению, как пустая доска (без фигур,
+// следующий ход первый).
+type Board struct {
```
по мере роста функциональности мне очень тяжело скроллить внутри одного файла туда-сюда. Имеет смысл в дальнейшем разбить на несколько файлов (например, отдельно структуры, отдельно методы доски move\get\put\remove, отдельно рокировки и enpassant) ?

nekr0z replied:
Меньше 1000 строк — это, вроде как, считается терпимо.
Но я согласен с критикой в том смысле, что у меня довольно хаотично методы по файлу распределились. Я их логично расставлю.
Вообще лично мне всегда удобнее стянуть ветку к себе на комп, и смотреть код в привычной IDE, а не в интерфейсе Github — но тут уж кому что нравится.

sadmadrus replied:
Тут согласен. Может вынести типы в отдельную папку models? вроде применяется такое порой

nekr0z replied:
Так некоторые делают, но вроде как в Go больше принято не отрывать методы от типов. Хотя, конечно, ситуации бывают разные.
Я переупорядочил и вынес всё, что про ходы, в отдельный файл, кажется, стало лучше.

nekr0z committed:
add pawn promotion

nekr0z committed:
reorganize code
